compiler/hash-types/src/new/pats.rs

```diff
@@ -2,12 +2,73 @@
 
 use hash_utils::{new_sequence_store_key, new_store, new_store_key, store::DefaultSequenceStore};
 
-// @@Todo
-#[derive(Debug, Clone, Copy)]
-pub enum Pat {}
+use super::{
+    control::{IfPat, OrPat},
+    data::CtorPat,
+    lits::LitPat,
+    scopes::BindingPat,
+    symbols::Symbol,
+    tuples::TuplePat,
+};
 
-new_sequence_store_key!(pub PatListId);
-pub type PatListStore = DefaultSequenceStore<PatListId, PatId>;
+/// A spread "pattern" (not part of [`Pat`]), which can appear in list patterns,
+/// tuple patterns, and constructor patterns.
+#[derive(Copy, Clone, Debug)]
+pub struct Spread {
+    /// The name of the spread bind.
+    /// If `name` does not map to a specific `Identifier` name, it means
+    /// that the bind is actually nameless.
+    pub name: Symbol,
+    /// The index in the sequence of target patterns, of this spread pattern.
+    pub index: usize,
+}
+
+/// A list pattern.
+///
+/// This is in the form `[x_1,...,x_n]`, with an optional spread `...(name?)` at
+/// some position.
+#[derive(Copy, Clone, Debug)]
+pub struct ListPat {
+    /// The sequence of patterns in the list pattern.
+    pub pats: PatListId,
+    /// The spread pattern, if any.
+    pub spread: Option<Spread>,
+}
+
+/// A range pattern containing two bounds `start` and `end`.
+///
+/// The `start` and `end` values must be either both [`LitPat::Int`], or both
+/// [`LitPat::Char`].
+#[derive(Copy, Clone, Debug)]
+pub struct RangePat {
+    /// The beginning of the range.
+    pub start: LitPat,
+    /// The end of the range.
+    pub end: LitPat,
+    /// If the range includes the `end` or not.
+    ///
+    /// If `true`, this is the pattern `start..<end`, otherwise it is
+    /// `start..end`.
+    pub end_exclusive: bool,
```

Comment:
I don't think `RangeEnd` should be replaced here with a `bool`. In this instance, I think verbosity is preferred because use cases of this fact become more self-documenting, especially in the exhaustiveness checking sources.

Reply:
That's fine, I don't think it is really that much of a difference. However I would want RangeEnd to not live in AST, but rather in source, same as you have done in your PR for int/float/etc literals.